### PR TITLE
statistic: fix panic when building topN

### PR DESCRIPTION
**Pull request by hi-rustin**
Signed-off-by: hi-rustin <rustin.liu@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/35948

Problem Summary:

### What is changed and how it works?
The issue arises when a value appears twice in the topN list, triggering a panic.  The solution involves breaking out of the loop as soon as we find the duplicate value. This bug seems to occur only when the column is of JSON type. Since we haven't been able to reproduce the issue, we've added a warning log to print all topN values, helping us identify the duplicated value.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] We will test it on users env.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fixed a bug that caused a panic while generating the topN statistics
```

### CODE REVIEW

**Review by hi-rustin — COMMENTED** [review]
🔢

**Review by winoros — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Comment by hi-rustin** [comment]
/hold

@time-and-fate Could you please take another look? Thanks!

**Review by time-and-fate — COMMENTED** [review]
1. Here we only logged the encoded bytes. But essentially, we need to know why two `types.Datum`s are encoded into the same bytes. If we only log the encoded bytes, we'll need some extra work to know that after we obtain the log.
I think it's best to log two `types.Datum`s with the same encoded bytes, with the full `Datum` struct printed. But that might not be easy to implement, so it's just a suggestion.
2. Have we verified that the bytes can be correctly escaped and printed by `zap`? Here the encoded bytes are probably not valid UTF-8 strings, and according to `zap` doc we should use `zap.Binary()` in this case.

File: pkg/statistics/builder.go [hunk 1]
```diff
@@ -373,12 +375,53 @@ func BuildHistAndTopN(
...
+							zap.String("category", "stats"),
+						).Error("try to convert datum to string failed", zap.Error(err))
+					}
+
+					logutil.BgLogger().With(
+						zap.String("category", "stats"),
+					).Warn(
+						"invalid sample data",
+						zap.Int64("i", i),
+						zap.Int64("j", int64(j)),
```
Comment (winoros) [inline pkg/statistics/builder.go:406]: we cannot use the `i` and `j` to do debugging?

File: pkg/statistics/builder.go [hunk 2]
```diff
@@ -373,12 +375,53 @@ func BuildHistAndTopN(
...
+						zap.String("datum", datumString),
+						zap.Binary("sampleBytes", sampleBytes),
+						zap.Binary("topNBytes", topNList[j].Encoded),
+						zap.String("firstTimeSample", firstTimeSampleString),
+					)
```
Comment (winoros) [inline pkg/statistics/builder.go:411]: We missed the `hist_id`(col/index's id) and `is_index`.

File: pkg/statistics/builder.go [hunk 3]
```diff
@@ -373,12 +375,53 @@ func BuildHistAndTopN(
...
+			foundTwice      bool
+			firstTimeSample types.Datum
+		)
 		for j := 0; j < len(topNList); j++ {
 			if bytes.Equal(sampleBytes, topNList[j].Encoded) {
-				// find the same value in topn: need to skip over this value in samples
+				// This should never happen, but we met this panic before, so we add this check here.
+				// See: https://github.com/pingcap/tidb/issues/35948
+				if foundTwice {
+					datumString, err := samples[i].Value.ToString()
```
Comment (time-and-fate) [inline pkg/statistics/builder.go:388]: If we meet this bug again, the `i` is actually `-1`, and I think we will panic again here.

File: pkg/statistics/builder.go (hunk 2 above)
Comment (hi-rustin) [reply pkg/statistics/builder.go:411]: Added.

**Review by winoros — APPROVED** [review]

File: pkg/statistics/builder.go (hunk 3 above)
Comment (hi-rustin) [reply pkg/statistics/builder.go]: Fixed.

**Review by time-and-fate — APPROVED** [review]

**Comment by hi-rustin** [comment]
Thanks for your review!  💚 💙 💜 💛 ❤️

**Pull request merged**